ocr/utils/sctk-2.4.10/doc/mergectm2rttm.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.29)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MERGECTM2RTTM 1"
.TH MERGECTM2RTTM 1 "2007-05-21" "perl v5.22.1" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
mergectm2rttm.pl \- Script that generate a RTTM file by merging CTM and RTTM.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&        mergectm2rttm.pl \-r rttmfile \-c ctmfile [ \-p pathtools ]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The script is a script which merged a ctm file and a rttm file to produce a rttm file. It takes in input a ctm file containing the lexemes information and a rttm file containing the speakers and segments information. It associates the lexemes and speakers present in the ctm according to the segments definitions from the rttm file (by segment definition, it means time intervals, filename and channel).
.SH "USAGE"
.IX Header "USAGE"
Examples:
.PP
.Vb 1
\&        align2html\-main.pl \-r file.rttm \-c file.ctm > mergedfile.rttm
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fBRequired file arguments:\fR" 4
.IX Item "Required file arguments:"
.RS 4
.PD 0
.IP "\fB\-r\fR, \fB\-\-rttmfile <rttmfile\fR>" 4
.IX Item "-r, --rttmfile <rttmfile>"
.PD
Path to the \s-1RTTM\s0 file.
.IP "\fB\-c\fR, \fB\-\-ctmfile <ctmfile\fR>" 4
.IX Item "-c, --ctmfile <ctmfile>"
Path to the \s-1CTM\s0 file.
.RE
.RS 4
.RE
.IP "\fBOptions:\fR" 4
.IX Item "Options:"
.RS 4
.PD 0
.IP "\fB\-p\fR,\fB\-\-pathtools <path\fR>" 4
.IX Item "-p,--pathtools <path>"
.PD
Path of the tools.
.RE
.RS 4
.RE
.IP "\fBOther options:\fR" 4
.IX Item "Other options:"
.RS 4
.PD 0
.IP "\fB\-h\fR, \fB\-\-help\fR" 4
.IX Item "-h, --help"
.PD
Display the help.
.RE
.RS 4
.RE
.SH "BUGS"
.IX Header "BUGS"
No known bugs.
.SH "NOTES"
.IX Header "NOTES"
As the script requires two other scripts: rttmSmooth.pl and rttmSort.pl, the path of the \s-1SCTK\s0 tools directory can be specified on the commandline using the '\-p' option.
.PP
The script checks if the two scripts are present and run the merging process.
To avoid any complication, there are 3 options to run the script:
(a) Run the script into its directory without the \-p option,
(b) Run the script from any location by specifying the path with \-p and the  path of the \s-1SCTK\s0 tools directory (.../stck/src/),
(c) Edit the mergectm2rttm.pl line at line 184 and 185 to give the locations of the two scripts.
.PP
The resulting rttm file is displayed in the standard output (\s-1STDOUT\s0), by doing the redirection '>', it can be stored into a file.
.SH "AUTHORS"
.IX Header "AUTHORS"
.IP "Jerome Ajot <jerome.ajot@nist.gov>" 4
.IX Item "Jerome Ajot <jerome.ajot@nist.gov>"
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
\&\s-1NIST\s0 Public Domain
.SH "DISCLAIMER"
.IX Header "DISCLAIMER"
This software was developed at the National Institute of Standards and Technology by employees of the Federal Government in the course of their official duties.  Pursuant to Title 17 Section 105 of the United States Code this software is not subject to copyright protection within the United States and is in the public domain. This software is an experimental system. \s-1NIST\s0 assumes no responsibility whatsoever for its use by any party.
.PP
\&\s-1THIS SOFTWARE IS PROVIDED \*(L"AS IS.\*(R"\s0 With regard to this software, \s-1NIST MAKES NO EXPRESS OR IMPLIED WARRANTY AS TO ANY MATTER WHATSOEVER, INCLUDING MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.\s0


ocr/utils/sctk-2.4.10/src/mergectm2rttm/mergectm2rttm.pl
#!/usr/bin/perl -w

# ALIGN2HTML
# Author: Jerome Ajot
#
# This software was developed at the National Institute of Standards and Technology by 
# employees of the Federal Government in the course of their official duties. Pursuant
# to title 17 Section 105 of the United States Code this software is not subject to
# copyright protection and is in the public domain. ALIGN2HTML is an experimental system.
# NIST assumes no responsibility whatsoever for its use by other parties, and makes no
# guarantees, expressed or implied, about its quality, reliability, or any other
# characteristic. We would appreciate acknowledgement if the software is used.
#
# THIS SOFTWARE IS PROVIDED "AS IS."  With regard to this software, NIST MAKES NO EXPRESS
# OR IMPLIED WARRANTY AS TO ANY MATTER WHATSOEVER, INCLUDING MERCHANTABILITY,
# OR FITNESS FOR A PARTICULAR PURPOSE.

use strict;
use Getopt::Long;
use Data::Dumper;

Getopt::Long::Configure(qw( auto_abbrev no_ignore_case ));

my $VERSION = "0.4";

my $RTTMFile = "";
my $CTMFile = "";
my $PATHTOOLS = "..";

sub usage
{
	print "perl mergectm2rttm.pl [OPTION] -r rttmfile -c ctmfile > outputfile\n";
	print "\n";
	print "Required file arguments:\n";
	print "  -r, --rttmfile           Path to the RTTM file.\n";
	print "  -c, --ctmfile            Path to the CTM file.\n";
	print "Path arguments:\n";
	print "  -p, --pathtools          Path of the tools.\n";
	print "\n";
}

sub loadRTTMFile
{
	my ($rttmFile, $rttmhash) = @_;
	
	open(RTTM, $rttmFile) or die "Unable to open for read RTTM file '$rttmFile'";
	
	while (<RTTM>)
	{
		chomp;
		
		# Remove comments which start with ;;
		s/;;.*$//;
		
		# Remove unwanted space at the begining and at the end of the line
		s/^\s*//;
		s/\s*$//;
		
		# if the line is empty then ignore
		next if ($_ =~ /^$/);
		
		my ($type, $file, $chnl, $tbeg, $tdur, $ortho, $stype, $spkrname, $conf) = split(/\s+/,$_,9);
		
		if(uc($type) eq "SPKR-INFO")
		{
		    $rttmhash->{SPKR}{$file}{$chnl}{$spkrname} = 1;
		}
		
		if(uc($type) eq "SPEAKER")
		{
			push( @{ $rttmhash->{DATA}{$file}{$chnl}{$tbeg} }, ($tbeg, $tdur, $ortho, $stype, $spkrname, $conf));
		}
	}
	
	close RTTM;
}

sub loadCTMFile
{
	my ($ctmFile, $ctmhash) = @_;
	
	open(CTM, $ctmFile) or die "Unable to open for read CTM file '$ctmFile'";
	
	while (<CTM>)
	{
		chomp;
		
		# Remove comments which start with ;;
		s/;;.*$//;
		
		# Remove unwanted space at the begining and at the end of the line
		s/^\s*//;
		s/\s*$//;
		
		# if the line is empty then ignore
		next if ($_ =~ /^$/);
		
		my ($file, $chnl, $tbeg, $tdur, $ortho, $conf, $stype, $spkrname) = split(/\s+/,$_,8);
		
		if($stype eq "lex")
		{
			my $mid = sprintf("%.4f", $tbeg+$tdur/2);
									
		    push( @{ $ctmhash->{$file}{$chnl}{$tbeg} }, ($tbeg, $tdur, $ortho, $stype, $conf, $mid));
		}
	}
	
	close CTM;
}

sub findSpeaker
{
	my ($rttmhash, $file, $chnl, $mid) = @_;
	my @listspkr;
	
	foreach my $tbeg(sort keys %{ $rttmhash->{DATA}{$file}{$chnl} })
	{
		my $bt = @{ $rttmhash->{DATA}{$file}{$chnl}{$tbeg} }[0];
		my $dur = @{ $rttmhash->{DATA}{$file}{$chnl}{$tbeg} }[1];
		my $spkrname = @{ $rttmhash->{DATA}{$file}{$chnl}{$tbeg} }[4];
		
		if( ($bt <= $mid) && ($mid <= sprintf("%.3f", $bt+$dur) ) )
		{
			push( @listspkr, [ ($spkrname, $tbeg) ] );
		}
	}
	
	if(scalar(@listspkr) == 0)
	{
		return undef;
	}
	else
	{
		return(\@{ $listspkr[int(rand(scalar(@listspkr)))] });
	}
}

sub MergedData
{
	my ($rttmhash, $ctmhash, $file, $chnl) = @_;
	
	if (exists($ctmhash->{$file}{$chnl}) )
	{
		foreach my $tbeg(sort {$a <=> $b} keys %{ $ctmhash->{$file}{$chnl} })
		{
			my $mid = $ctmhash->{$file}{$chnl}{$tbeg}[5];
		    my $spkrtbeg = findSpeaker($rttmhash, $file, $chnl, $mid);
		    
		    if($spkrtbeg)
		    {
		    	my $ctmbeg = $ctmhash->{$file}{$chnl}{$tbeg}[0];
		    	my $ctmdur = $ctmhash->{$file}{$chnl}{$tbeg}[1];
		    	my $ctmortho = $ctmhash->{$file}{$chnl}{$tbeg}[2];
		    	my $ctmstype = $ctmhash->{$file}{$chnl}{$tbeg}[3];
		    	my $ctmconf = $ctmhash->{$file}{$chnl}{$tbeg}[4];
		    	my $ctmspkr = $spkrtbeg->[0];
		    	
		    	print OUTPUT "SPEAKER $file $chnl $ctmbeg $ctmdur <NA> <NA> $ctmspkr <NA>\n";
		    	print OUTPUT "LEXEME $file $chnl $ctmbeg $ctmdur $ctmortho $ctmstype $ctmspkr $ctmconf\n";
		    }
		}
	}
}

GetOptions
(
    'rttmfile=s' => \$RTTMFile,
    'ctmfile=s'  => \$CTMFile,
    'pathtools=s'  => \$PATHTOOLS,
    'version'    => sub { print "mergectm2rttm version: $VERSION\n"; exit },
    'help'       => sub { usage (); exit },
);

die "ERROR: An RTTM file must be set." if($RTTMFile eq "");
die "ERROR: An CTM file must be set." if($CTMFile eq "");

my %RTTM;
my %CTM;

loadRTTMFile($RTTMFile, \%RTTM);
loadCTMFile($CTMFile, \%CTM);

my $RTTMSMOOTH = "$PATHTOOLS/rttmSmooth/rttmSmooth.pl";
my $RTTMSORT = "$PATHTOOLS/rttmSort/rttmSort.pl";

die "ERROR: 'rttmSmooth.pl' cannot be found, please specify the tools path with -p\n" if(! -e $RTTMSMOOTH);
die "ERROR: 'rttmSort.pl' cannot be found, please specify the tools path with -p\n" if(! -e $RTTMSORT);

open(OUTPUT, "| $RTTMSMOOTH | $RTTMSORT");

print ";; type file chnl tbeg tdur ortho stype name conf\n";

foreach my $file(sort keys %{ $RTTM{SPKR} })
{
	foreach my $chnl(sort keys %{ $RTTM{SPKR}{$file} })
	{
		foreach my $spkrname(sort keys %{ $RTTM{SPKR}{$file}{$chnl} })
		{
			print OUTPUT "SPKR-INFO $file $chnl <NA> <NA> <NA> unknown $spkrname <NA>\n";
		}
		
		MergedData(\%RTTM, \%CTM, $file, $chnl);
	}
}

close(OUTPUT);

